src/main/java/uk/ac/ebi/ampt2d/accession/service/BlockState.java

```diff
@@ -0,0 +1,172 @@
+/*
+ *
+ * Copyright 2018 EMBL - European Bioinformatics Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package uk.ac.ebi.ampt2d.accession.service;
+
+import org.springframework.data.util.Pair;
+import uk.ac.ebi.ampt2d.accession.serial.block.MonotonicRange;
+import uk.ac.ebi.ampt2d.accession.serial.block.MonotonicRangePriorityQueue;
+import uk.ac.ebi.ampt2d.accession.serial.block.persistence.entities.ContiguousIdBlock;
+import uk.ac.ebi.ampt2d.accession.service.exceptions.AccessionIsNotPending;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+/**
+ * This class holds the state of the monotonic id blocks used at this moment on the application.
+ * This class is not thread safe.
+ */
+class BlockState {
+
+    private final PriorityQueue<ContiguousIdBlock> assignedBlocks;
+
+    private final MonotonicRangePriorityQueue availableRanges;
+
+    private final HashSet<Long> generatedAccessions;
+
+    private final PriorityQueue<Long> committedAccessions;
+
+    public BlockState() {
+        this.assignedBlocks = new PriorityQueue<>(ContiguousIdBlock::compareTo);
+        this.availableRanges = new MonotonicRangePriorityQueue();
+        this.generatedAccessions = new HashSet<>();
+        this.committedAccessions = new PriorityQueue<>(Long::compareTo);
+    }
+
+    public void addBlock(ContiguousIdBlock block) {
+        assignedBlocks.add(block);
+        availableRanges.add(new MonotonicRange(block.getLastCommitted() + 1, block.getEnd()));
+    }
+
+    public MonotonicRangePriorityQueue getAvailableRanges() {
+        return availableRanges;
+    }
+
+    /**
+     * Polls the next continuous array of monotonic values.
+     *
+     * @param maxValues Max array size returned by the function
+     * @return
+     */
+    public long[] pollNextMonotonicValues(int maxValues) {
```

Comment:
I think in this case `pollNext` is good because it provides the exact meaning of what the values are. The implementation detail that they are stored in a queue shouldn't be leaked to the interface and its consumers.